chore: Add Prettier for Code Formatting

MrArnoldPalmer opened the pull request:
Adds prettier to auto-format code. Also adds eslint-plugin-prettier to
embed formatting into the lint workflow. Running `yarn lint` will report
any formatting mis-match as a lint error, `yarn lint:fix` will fix any
formatting related issues.

We are pretty close to the prettier defaults, if we cared to not have any extra configuration we could delete the .prettierrc.yaml file and use doublequotes for strings as the default. I don't really care either way personally so continuing with singlequotes just means less of a diff. This also removes all of the style-only related lint rules from our eslint config just to trim it down.

This obviously adds a lot of noise to the git history and can affect the usefulness of git blame. Just something to consider.
<!--
Explain what changed and why.

Please read the [Contribution guidelines][1] and follow the pull-request
checklist.

[1]: https://github.com/aws/jsii/blob/master/CONTRIBUTING.md
-->

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-Blkkw9bQFn8A
* Commit ID: 341ded13769ee9b5c46f6dc3a613060ced9103e7
* Result: SUCCEEDED
* [Build Logs](https://d445wms7qi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9d8372a4-94c6-4bbb-94d4-d3834e1cafc1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rix0rrr commented:
Although I'm personally not sure we need this (I don't feel like this is solving a problem I personally ever had, though I can't speak for other people obviously) I'm provisionally okay with this, but I must have IDE support for this otherwise it's going to cause grief to no end when my build fails for trivial reasons after I'm done typing. Can you write a little something in the CONTRIBUTING guide on how to configure inline feedback in the IDE?

Style-wise, not too keen on all the newlines it's adding everywhere, but I guess that's the price we pay...

RomainMuller commented on packages/@jsii/integ-test/test/build-cdk.test.ts:
```diff
@@ -27,62 +36,98 @@ describe('Build CDK', () => {
+          'awslint',
+          'run',
+          'build'
+        ],
```
I don't like this change :/ It hurts readability :/

rix0rrr replied:
I agree. Too many newlines.

RomainMuller commented on packages/@jsii/kernel/lib/api.ts:
```diff
@@ -220,44 +220,46 @@ export interface NamingRequest {
-  readonly naming: { readonly [language: string]: { readonly [key: string]: any } | undefined };
+  readonly naming: {
+    readonly [language: string]: { readonly [key: string]: any } | undefined;
+  };
```
I find it odd that it applies to `[lamguage: string]` but not to `[key: string]`... I feel like they should be treated the same.

skinny85 commented:
I don't like the fact it removes trailing commas (not using trailing commas makes diffs look bad) :/. Not sure if this is a matter of configuration, or if prettier doesn't support them.

MrArnoldPalmer commented:
@rix0rrr Yeah in JSII I don't think this is really solving a lot of problems that we currently have. I think it would have more of an impact in aws-cdk if we decided we like this toolchain and want to add it there. I did this now after talking with @nija-at about inconsistencies in newlines at EOF in aws-cdk and possible fixes.

I agree that linebreaks are not ideal in some places. Prettier allows ignoring a node in the AST with `// prettier-ignore`. Added this on spawn `[...args]` to demonstrate.

Added notes about IDE tooling, but the long term idea is to add a github action that will run formatting on PRs and re-commit automatically. Then users can write new code however they prefer and it just gets normalized as it moves to master.

@skinny85 I agree, I changed the prettier "trailingComma" option to "all" instead of default which is "es5".

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-Blkkw9bQFn8A
* Commit ID: 15b46aa7edeaa67dcfc254277dad0f57c53eaca1
* Result: SUCCEEDED
* [Build Logs](https://d445wms7qi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e4ef526a-465f-4050-974e-b841d2b86d49%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-Blkkw9bQFn8A
* Commit ID: d95f33f7932de8d45b4c8672e75cfc0074028e28
* Result: SUCCEEDED
* [Build Logs](https://d445wms7qi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f713349a-c108-4b87-817b-dd47f06bd3c5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

RomainMuller commented:
This looks like it got severely merge-conflicted. Would you mind rebasing/re-prettiftying so we can merge?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-Blkkw9bQFn8A
* Commit ID: 8e79cdea0fa735e1182669e93f54fa27b546c436
* Result: SUCCEEDED
* [Build Logs](https://d445wms7qi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6c7cbd71-62ad-475a-b192-f0904261e60c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-Blkkw9bQFn8A
* Commit ID: e2bb1f1067672e6a143859c2ea62f7f1a0d69b6a
* Result: SUCCEEDED
* [Build Logs](https://d445wms7qi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2efcfb08-9254-4fbd-a73b-e4dffbc06030%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-Blkkw9bQFn8A
* Commit ID: f5ce5ff2274fb5ac23c89bff4262a53b03649258
* Result: SUCCEEDED
* [Build Logs](https://d445wms7qi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4be01230-35f6-4afc-aa72-d34bb84f4d0e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

MrArnoldPalmer commented:
@RomainMuller dismissed review because I upgraded prettier to 2.0 which has better support for complex types in TS so some of the outputs have changed. This was required for TS 3.3 + private fields support.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-Blkkw9bQFn8A
* Commit ID: e194a3ffada0b2cad8d7d99e517faec429dc7883
* Result: SUCCEEDED
* [Build Logs](https://d445wms7qi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2a84eb57-3e05-4a49-84b6-4fc963574f72%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-Blkkw9bQFn8A
* Commit ID: c11fbfa2be397f3156b0468531d9ad1c07e529eb
* Result: SUCCEEDED
* [Build Logs](https://d445wms7qi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c7f93ed1-3cae-4cd4-afc8-5354bf389844%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-Blkkw9bQFn8A
* Commit ID: 268f60cf895d9960cdb0c19f8c1cb8fd79c4adba
* Result: SUCCEEDED
* [Build Logs](https://d445wms7qi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=abb0c154-d1b7-4fae-83de-39cf28813697%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
